server/src/main/java/org/elasticsearch/action/search/SearchResponse.java

```diff
@@ -614,18 +682,246 @@ public int hashCode() {
+            this.skippedShards = skippedShards;
+            this.failedShards = failedShards;
+            this.failures = failures == null ? Collections.emptyList() : Collections.unmodifiableList(failures);
+            this.took = took;
+            this.timedOut = timedOut;
+        }
+
+        public Cluster(StreamInput in) throws IOException {
+            this.clusterAlias = in.readString();
+            this.indexExpression = in.readString();
+            this.status = Status.valueOf(in.readString().toUpperCase(Locale.ROOT));
+            this.totalShards = in.readOptionalInt();
+            this.successfulShards = in.readOptionalInt();
+            this.skippedShards = in.readOptionalInt();
+            this.failedShards = in.readOptionalInt();
+            Long took = in.readOptionalLong();
+            if (took == null) {
+                this.took = null;
+            } else {
+                this.took = new TimeValue(took);
+            }
+            this.timedOut = in.readBoolean();
+            this.failures = Collections.unmodifiableList(in.readList(ShardSearchFailure::readShardSearchFailure));
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeString(clusterAlias);
+            out.writeString(indexExpression);
+            out.writeString(status.toString());
+            out.writeOptionalInt(totalShards);
+            out.writeOptionalInt(successfulShards);
+            out.writeOptionalInt(skippedShards);
+            out.writeOptionalInt(failedShards);
+            out.writeOptionalLong(took == null ? null : took.millis());
+            out.writeBoolean(timedOut);
+            out.writeList(failures);
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            String name = clusterAlias;
+            if (clusterAlias.equals("")) {
+                name = "(local)";
+            }
+            builder.startObject(name);
+            {
+                builder.field("status", getStatus().toString());
+                builder.field("indices", indexExpression);
+                if (took != null) {
+                    builder.field("took", took.millis());
+                }
+                builder.field("timed_out", timedOut);
+                if (totalShards != null) {
+                    builder.startObject("_shards");
+                    builder.field("total", totalShards);
+                    if (successfulShards != null) {
+                        builder.field("successful", successfulShards);
+                    }
+                    if (skippedShards != null) {
+                        builder.field("skipped", skippedShards);
+                    }
+                    if (failedShards != null) {
+                        builder.field("failed", failedShards);
+                    }
+                    builder.endObject();
```

Comment:
I'm not sure what you are asking/proposing. Do you mean that I should move this "toXContent" functionality up to the `Clusters.toXContent`? 

That's where this one is called from. The Clusters.toXContent calls each individual `Cluster.toXContent`:

```
        @Override
        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
            if (total > 0) {
                builder.startObject(_CLUSTERS_FIELD.getPreferredName());
                builder.field(TOTAL_FIELD.getPreferredName(), total);
                builder.field(SUCCESSFUL_FIELD.getPreferredName(), getSuccessful());
                builder.field(SKIPPED_FIELD.getPreferredName(), getSkipped());
                // TODO: add FAILED_FIELD
                if (clusterInfo.size() > 0) {
                    builder.startObject("details");
                    for (AtomicReference<Cluster> cluster : clusterInfo.values()) {
                        cluster.get().toXContent(builder, params);   // call to each Cluster.toXContent here
                    }
                    builder.endObject();
                }
                builder.endObject();
            }
            return builder;
        }
```

The model I used here for the REST layer "serialization" is patterned after how we handle the Transport layer serialization, where `Clusters.writeTo` (and read) relies on each Cluster to do its portion of the Writeable action. Here is the `Clusters.writeTo` method:

```
        @Override
        public void writeTo(StreamOutput out) throws IOException {
            out.writeVInt(total);
            out.writeVInt(successful);
            out.writeVInt(skipped);
            if (out.getTransportVersion().onOrAfter(TransportVersion.V_8_500_052)) {
                if (clusterInfo != null) {
                    List<Cluster> clusterList = clusterInfo.values().stream().map(AtomicReference::get).toList();
                    out.writeList(clusterList);
                } else {
                    out.writeList(Collections.emptyList());
                }
            }
        }
```

Sorry if I missed what you are suggesting. Let me know what you'd prefer.

Reply:
it's fine and not so important. I was referring to the fact that search already outputs a _shards section that looks the same as this one, and I wondered if those two toXcontent could be unified. It won't save a lot of code, but it would make it clear that the two are aligned in terms of naming for a good reason. This could totally be a follow-up.

Reply:
Oh, I see. I can try that out in the next ticket (I have two or three follow-ons to this one).